multi: Add Balancer for round-robin redundancy

maicalal opened the pull request:
Added a RoundRobinBalancer to do a round robin over a list of 
hosts using a picker. List of hosts to be set in connection config 
thats used to create a new Client.
Changes in file roundrobin.go:
- Picker
    - responsible to pick the next usable connection 

- RoundrobinBalancer
    - holds a picker instance to get the next connection
    - maintains a map for storing the current connection 
       state for a host
    - maintains a map of socket connections by host

Changes in file infrastructure.go:
- Client now has a reference to roundrobinbalancer
- Changes to following workflows
    - Creating New client
        - Does a rrrb.Get() to have one ready connection
    - wsInHandler
        - iterates over all active connections to check for
           messages
    - wsOutHandler
        - picks a connection using balancer to send out 
          message
    - wsReconnectHandler
        - gets the disconnected connections from balancer
          and attempts to reconnect.
          If all were disconnected, it would flag to call start()
This fixes #884

davecgh commented on rpcclient/infrastructure.go:
```diff
@@ -126,7 +126,7 @@ type Client struct {
 	// wsConn is the underlying websocket connection when not in HTTP POST
 	// mode.
-	wsConn *websocket.Conn
+	// wsConn *websocket.Conn
```
Please remove this and the associated comment.

davecgh commented on rpcclient/infrastructure.go:
```diff
@@ -136,11 +136,7 @@ type Client struct {
-	// retryCount holds the number of times the client has tried to
-	// reconnect to the RPC server.
-	retryCount int64
+	//disconnected bool
```
Same here.

davecgh commented on rpcclient/infrastructure.go:
```diff
@@ -159,6 +155,9 @@ type Client struct {
 	shutdown        chan struct{}
 	wg              sync.WaitGroup
+
+	//balancer
```
Please describe what this does better.  This comment isn't useful.

davecgh commented on rpcclient/infrastructure.go:
```diff
@@ -412,17 +411,27 @@ out:
-			if c.shouldLogReadError(err) {
-				log.Errorf("Websocket receive error from "+
-					"%s: %v", c.config.Host, err)
+		//check for all wsConns with balancer
```
Please follow the conventions of the other comments in the code.  e.g. full sentences with a period and a space after the double slashes.

This is the true throughout the entire PR, but I'll refrain from commenting on each instance and just mention it here.

davecgh commented on rpcclient/infrastructure.go:
```diff
@@ -1242,18 +1280,16 @@ func New(config *ConnConfig, ntfnHandlers *NotificationHandlers) (*Client, error
-		config:          config,
-		wsConn:          wsConn,
+		config: config,
+		//wsConn:          wsConn,
```
Remove.

davecgh commented on rpcclient/roundrobin.go:
```diff
@@ -0,0 +1,347 @@
+
+//GetNextDisconnectedWsConn will iterate over the connection state map and
+//return the first wsConnection that has its state as Disconnected
+func (rrb *RoundRobinBalancer) GetNextDisconnectedWsConn() *HostAddress {
```
The names of functions that are designed as getters should not be named with `Get` per [Effective Go Guidelines](https://golang.org/doc/effective_go.html#Getters) which is part of the [Code Contribution Guidelines](https://github.com/decred/dcrd/blob/master/docs/code_contribution_guidelines.md#ReqReading).

It should be `NextDisconnectedWsConn`.

davecgh commented on rpcclient/roundrobin.go:
```diff
@@ -0,0 +1,347 @@
+}
+
+//GetConnectionState will return the connection state for given host
+func (rrb *RoundRobinBalancer) GetConnectionState(host string) (state ConnectionState, ok bool) {
```
Same here regrading `Get`.

davecgh commented on rpcclient/roundrobin.go:
```diff
@@ -0,0 +1,347 @@
+}
+
+// GetWsConnection gets the wsConn corresponding to the host
+func (rrb *RoundRobinBalancer) GetWsConnection(host string) (wsConn *websocket.Conn, connOk bool) {
```
Same here regrading `Get`.

davecgh commented on rpcclient/roundrobin.go:
```diff
@@ -0,0 +1,347 @@
+//If no items in HostAddresses list, then would add one using *ConnConfig.Host
+//else use the *ConnConfig.HostAddresses
+func (cn *Client) BuildBalancer(cc *ConnConfig) *RoundRobinBalancer {
+
```
Superfluous newline.

maicalal replied:
Removed extra lines.

davecgh commented on rpcclient/roundrobin.go:
```diff
@@ -0,0 +1,347 @@
+//Get gets the next connection to be used.
+//It returns both *websocket.Conn and *HostAddress
+func (rrb *RoundRobinBalancer) Get() (*websocket.Conn, *HostAddress, error) {
+
```
Superfluous newline.

maicalal replied:
Removed extra lines.

davecgh commented on rpcclient/roundrobin.go:
```diff
@@ -0,0 +1,347 @@
+
+//Get gets the next connection to be used.
+//It returns both *websocket.Conn and *HostAddress
+func (rrb *RoundRobinBalancer) Get() (*websocket.Conn, *HostAddress, error) {
```
I would call this `NextWsConn` since it is more descriptive and doesn't run afoul of the effective Go guidelines mentioned further down in the review.

maicalal replied:
Renamed it to NextConn since wsConn may be nil in case of HttpPostMode.

davecgh commented:
Just an FYI that I will be doing some thorough testing and review on this after the upcoming release.  While we always try to make sure everything is well tested and functioning before it goes to master, this is double the case when a release is looming.

maicalal commented:
Thanks @davecgh , will wait for the feedback.

maicalal commented:
hi @davecgh any luck :) ...

davecgh commented on rpcclient/infrastructure.go:
```diff
@@ -159,6 +148,13 @@ type Client struct {
+	// Whenever we need a connection, RoundRobinBalancer.NextConn() should be used to
+	// retrive a websocket connection or a HostAddress in case of HttpPostMode.
+	// It relies on the HostAddresses present in ConnConfig as the source for round-robin.
+	rrbalancer *RoundRobinBalancer
```
I would suggest just calling this `balancer` so that it can potentially provide other load balancing methods in the future.

davecgh commented on rpcclient/infrastructure.go:
```diff
@@ -412,17 +408,29 @@ out:
+		// To check for incoming messages, iterate over the HostAddresses
+		// present with the config. Check if its an active connection
+		// and go ahead with ReadMessage.
+		for i := 0; i < len(c.config.HostAddresses); i++ {
```
This doesn't seem like a very efficient way to handle multiple connections.  I would highly suggest just launching a separate `wsInHandler` goroutine per connection that is established instead.  The logic is already based around channels for concurrency.

e.g.
```Go
func (c *Client) wsInHandler(wsConn *websocket.Conn) {
...

		_, msg, err := c.wsConn.ReadMessage()
		if err != nil {
			if c.shouldLogReadError(err) { ... }
			c.balancer.Close(wsConn)
			break out
		}
		c.handleMessage(msg)
...
}
```

Notice that I also changed the `Close` method on the balancer to take the websocket connection which is more robust than dealing with hostnames.  That method would need to be updated accordingly.

Also, I would suggest instead of having `Close` treat an empty string as the disconnect all logic, create a separate method such as `CloseAll` which simply loops the map of open websocket connections and calls `Close` on each one.

davecgh commented on rpcclient/roundrobin.go:
```diff
@@ -0,0 +1,367 @@
+// Copyright (c) 2015-2017 The Decred developers
```
2018

davecgh commented on rpcclient/roundrobin.go:
```diff
@@ -0,0 +1,367 @@
+var (
+	// ErrNoConnAvailable indicates no Conn is available for pick().
+	ErrNoConnAvailable = errors.New("no Connection is available")
+	// ErrNoUsableConnAvailable indicates no usable connections available i.e.
```
Need a blank line before for code consistency.

maicalal replied:
Fixed it.

davecgh commented on rpcclient/roundrobin.go:
```diff
@@ -0,0 +1,367 @@
+// ConnectionState indicates the current state of connection.
+type ConnectionState int
+
+func (s ConnectionState) String() string {
```
Missing function comment.

maicalal replied:
Fixed it.

davecgh commented on rpcclient/roundrobin.go:
```diff
@@ -0,0 +1,367 @@
+)
+
+// ConnectionState indicates the current state of connection.
+type ConnectionState int
```
None of this is used outside of the `rpcclient`, so it should all be unexported.

maicalal replied:
Now since its used in balancer interface, we need to export it.

davecgh commented on rpcclient/infrastructure.go:
```diff
@@ -412,17 +408,29 @@ out:
+					}
+					c.rrbalancer.Close(hostAdd.Host)
+					if c.rrbalancer.IsAllDisconnected() {
+						c.rrbalancer.NeedsClientRestart = true
```
I think this should be handled by the balancer `Close` method.

davecgh commented on rpcclient/roundrobin.go:
```diff
@@ -0,0 +1,367 @@
+const (
+	// Idle indicates the conn is idle.
+	Idle ConnectionState = iota
+	// Connecting indicates the conn is connecting.
```
Need blank lines between each def for code consistency.

davecgh commented on rpcclient/roundrobin.go:
```diff
@@ -0,0 +1,367 @@
+type Balancer interface {
+	// NextConn gets the next usabel connection.
+	NextConn() (wsConn *websocket.Conn, hostAddress *HostAddress, err error)
+	// ConnectionState gets the connectionstate of corresponding wsConn.
```
Same here regarding the blank lines.

davecgh commented:
Please either leave a comment on the PR or ping me via one of the chat apps once you've updated as github isn't the best about notifying on updates to PRs otherwise.

maicalal commented:
Hi @davecgh ,
Comments got screwed up in my last checkin. Here is the same:
-------------------------------------------
Incorporating davecgh’s feedback
- WsInHandler to be per wsConn
- Moved the Check and setting flag to indicate need for client restart to balancer.close()
- Added balancer.CloseAll()
- Other formatting changes
-------------------------------------------

davecgh commented on rpcclient/roundrobin.go:
```diff
@@ -0,0 +1,434 @@
+
+var (
+	// ErrNoConnAvailable indicates no Conn is available for pick().
+	ErrNoConnAvailable = errors.New("no Connection is available")
```
The word connection shouldn't be capitalized.  It is not good practice to refer to the actual struct names in error messages.

davecgh commented on rpcclient/roundrobin.go:
```diff
@@ -0,0 +1,434 @@
+
+	// ErrNoUsableConnAvailable indicates no usable connections available i.e.
+	// client is marked as Shutdown.
+	ErrNoUsableConnAvailable = errors.New("no Connection is available for use, all are marked as Shutdown")
```
Same here for but the struct name and the connection state.  Also, it's not a grammatically correct error since it contains a comma splice.

I would suggest either: 

```Go
ErrNoUsableConnAvailable = errors.New("no connections are available for use")
```
or 

```Go
ErrNoUsableConnAvailable = errors.New("all connections have been shutdown")
```

davecgh commented on rpcclient/roundrobin.go:
```diff
@@ -0,0 +1,434 @@
+
+// Balancer uses Picker to pick connections.
+// It collects and maintains the connectivity states.
+type Balancer interface {
```
This should not be exported.  This interface is not directly usable by callers outside of the `rpcclient` package and it pollutes the public API by being exported.

davecgh commented on rpcclient/roundrobin.go:
```diff
@@ -0,0 +1,434 @@
+)
+
+// HostAddress represents a server the client connects to.
+type HostAddress struct {
```
This should not be exported.  This struct is not directly usable by callers outside of the `rpcclient` package and it pollutes the public API by being exported.

maicalal replied:
Client would need to pass array of HostAddress as part of the config.
For the same, we would need to export HostAddress.

davecgh commented on rpcclient/roundrobin.go:
```diff
@@ -0,0 +1,434 @@
+
+// WsInHandlerForConn represents the incoming message handler for a
+// ws connection. This is invoked as a goroutine, one per ws connection
+type WsInHandlerForConn func(*websocket.Conn, string)
```
This should not be exported.  It is not directly usable by callers outside of the `rpcclient` package and it pollutes the public API by being exported.

davecgh commented on rpcclient/roundrobin.go:
```diff
@@ -0,0 +1,434 @@
+
+// RoundRobinBalancer represents a round-robin balancer. It implements Balancer
+// interface and maintains a Picker for picking Hosts in a round-robin manner.
+type RoundRobinBalancer struct {
```
This should not be exported.  It is not directly usable by callers outside of the `rpcclient` package and it pollutes the public API by being exported.

davecgh commented on rpcclient/roundrobin.go:
```diff
@@ -0,0 +1,434 @@
+}
+
+// Picker is used to pick the next connection to be used.
+type Picker interface {
```
This should not be exported.  It is not directly usable by callers outside of the `rpcclient` package and it pollutes the public API by being exported.

davecgh commented on rpcclient/roundrobin.go:
```diff
@@ -0,0 +1,434 @@
+// BuildBalancer is used to setup balancer with required config.
+// If no items in HostAddresses list, then would add one using *ConnConfig.Host
+// and *ConnConfig.Endpoint else use the *ConnConfig.HostAddresses for round-robin.
+func (cn *Client) BuildBalancer(cc *ConnConfig) *RoundRobinBalancer {
```
Please don't use variable names that are inconsistent with the rest of the package. The receiver for the *Client is `c` in every other instance in the package.  Also, the parameter name for *ConnConfig is `config` in every other instance in the package.

maicalal replied:
This, have resolved in my latest changes.

davecgh commented on rpcclient/roundrobin.go:
```diff
@@ -0,0 +1,434 @@
+// If no items in HostAddresses list, then would add one using *ConnConfig.Host
+// and *ConnConfig.Endpoint else use the *ConnConfig.HostAddresses for round-robin.
+func (cn *Client) BuildBalancer(cc *ConnConfig) *RoundRobinBalancer {
+	if len(cc.HostAddresses) <= 0 {
```
The length of a slice can never be negative.  Should be `== 0`.

davecgh commented on rpcclient/infrastructure.go:
```diff
@@ -458,11 +455,20 @@ out:
-				c.Disconnect()
+				// Failed to get any usable connection, lets break out.
+				log.Tracef("RPC client balancer error: %s", err)
+				c.balancer.NeedsClientRestart = true
```
This assumes that `c.balancer` is specifically the concrete type of `RoundRobinBalance`.  It appears the intent was to have a `Balancer` interface which the rest of the code would be agnostic to.

This logic doesn't seem quite right to achieve that goal.

davecgh commented on rpcclient/infrastructure.go:
```diff
@@ -613,7 +619,7 @@ func (c *Client) resendRequests() {
 		// Stop resending commands if the client disconnected again
 		// since the next reconnect will handle them.
-		if c.Disconnected() {
+		if c.balancer.IsAllDisconnected() {
```
This is only defined on `RoundRobinBalancer`, so as mentioned elsewhere, this breaks the notion of the code being agnostic to the type of balancer being used.

davecgh commented on rpcclient/infrastructure.go:
```diff
@@ -1304,20 +1345,19 @@ func (c *Client) Connect(ctx context.Context, retry bool) error {
 		return ErrNotWebsocketClient
 	}
-	if c.wsConn != nil {
+	if c.balancer.isReady {
```
This is only defined on `RoundRobinBalancer`, so as mentioned elsewhere, this breaks the notion of the code being agnostic to the type of balancer being used.

jrick commented:
This change strikes me as being backwards.  I would expect a load balances to be created from many clients, and not a single client load balancing over multiple servers.  It should be possible to create a load balancing client in another package using only the exported features of the rpcclient package.  Creating a new type to handle load balanced requests also means we can remove the methods that keep state and can't be load balanced.  Is there some reason I'm missing why it wasn't designed this way?

maicalal commented:
Hi @jrick this was the base requirement for issue #884 .
It made sense to have a pure client side balancer where it has multiple servers that client can connect to, so went ahead with the changes.
Moreover, a load balancer over multiple clients, would change its meaning and would be more like a message broker in a pub/sub paradigm.

dnldd commented:
Testing. @maicalal please resolve the remaining review issues.

dnldd commented on rpcclient/balancer.go:
```diff
@@ -0,0 +1,67 @@
+// Balancer defines requirements for a load balancer
+// using list of hostaddresses as the source.
+type Balancer interface {
+	// NextConn gets the next usabel connection.
```
s/usabel/usable

dnldd commented:
When the client fails to connect to all host addresses it currently only retries the last host address provided. I think it should cycle through all provided hosts instead.
```
2018-09-10 15:37:12.888 [INF] RPCC: Balancer: Failed to connect to localhost:19110: dial tcp [::1]:19110: connect: connection refused
2018-09-10 15:37:12.891 [INF] RPCC: Balancer: Failed to connect to localhost:19111: dial tcp [::1]:19111: connect: connection refused
2018-09-10 15:37:12.894 [INF] RPCC: Balancer: Failed to connect to localhost:19109: dial tcp [::1]:19109: connect: connection refused
2018-09-10 15:37:12.896 [INF] RPCC: Balancer: Failed to connect to localhost:19109: dial tcp [::1]:19109: connect: connection refused
2018-09-10 15:37:12.899 [INF] RPCC: Balancer: Failed to connect to localhost:19109: dial tcp [::1]:19109: connect: connection refused
2018-09-10 15:37:12.901 [INF] RPCC: Balancer: Failed to connect to localhost:19109: dial tcp [::1]:19109: connect: connection refused
2018-09-10 15:37:12.904 [INF] RPCC: Balancer: Failed to connect to localhost:19109: dial tcp [::1]:19109: connect: connection refused
...
```

dnldd commented:
I'm expecting the client to start cycling through provided hosts for next viable connection when the connection drops for the current host.

```
2018-09-10 16:06:05.972 [INF] RPCC: Balancer: Failed to connect to localhost:19110: dial tcp [::1]:19110: connect: connection refused
2018-09-10 16:06:05.975 [INF] RPCC: Balancer: Failed to connect to localhost:19111: dial tcp [::1]:19111: connect: connection refused
2018-09-10 16:06:06.207 [INF] RPCC: Balancer: Established connection to RPC server localhost:19109
2018-09-10 16:06:06.207 [INF] RPCC: Balancer: Connection pick for RPC localhost:19109
2018-09-10 16:06:06.208 [INF] RPCC: Balancer: Connection pick for RPC localhost:19109
2018-09-10 16:06:06.208 [INF] RPCC: Balancer: Will be using localhost:19109 for notification: notifyblocks
2018/09/10 16:06:06 NotifyBlocks: Registration Complete
2018-09-10 16:06:11.213 [INF] RPCC: Balancer: Connection pick for RPC localhost:19109
2018/09/10 16:06:11 Block count: 23512
2018-09-10 16:06:16.212 [INF] RPCC: Balancer: Connection pick for RPC localhost:19109
2018/09/10 16:06:16 Block count: 23512
2018-09-10 16:06:21.213 [INF] RPCC: Balancer: Connection pick for RPC localhost:19109
2018/09/10 16:06:21 Block count: 23512
2018-09-10 16:06:26.214 [INF] RPCC: Balancer: Connection pick for RPC localhost:19109
2018/09/10 16:06:26 Block count: 23512
2018-09-10 16:06:31.214 [INF] RPCC: Balancer: Connection pick for RPC localhost:19109
2018/09/10 16:06:31 Block count: 23512
2018-09-10 16:06:36.214 [INF] RPCC: Balancer: Connection pick for RPC localhost:19109
2018/09/10 16:06:36 Block count: 23512
2018-09-10 16:06:41.216 [INF] RPCC: Balancer: Connection pick for RPC localhost:19109
2018/09/10 16:06:41 Block count: 23512
2018-09-10 16:06:46.215 [INF] RPCC: Balancer: Connection pick for RPC localhost:19109
2018/09/10 16:06:46 Block count: 23512
2018-09-10 16:06:51.214 [INF] RPCC: Balancer: Connection pick for RPC localhost:19109
2018/09/10 16:06:51 Block count: 23512
2018-09-10 16:06:56.215 [INF] RPCC: Balancer: Connection pick for RPC localhost:19109
2018/09/10 16:06:56 Block count: 23512
2018-09-10 16:06:57.859 [ERR] RPCC: Websocket receive error from localhost:19109: websocket: close 1006 (abnormal closure): unexpected EOF
2018-09-10 16:07:01.216 [INF] RPCC: Balancer: Connection pick for RPC localhost:19109
2018/09/10 16:10:49 the client has been shutdown
```

maicalal commented:
Thanks @dnldd working on it...

maicalal commented:
Hi @dnldd , have fixed both the issues. Please try out with latest.
Earlier, code was assuming connection is down, when it failed for the first time.
Now, it would round-robin for initial connection as well. 
Have put a limit of 50 trials for a host, for the initial connection, i.e. the first time a host gets successfully connected.

maicalal commented:
Apologies, for the last checkin got a little messy. Will fix it during final squash.

dnldd commented:
There are still review points by @davecgh yet to be resolved. Currently the balancer tries to establish connections for all the hosts even after establishing a connection for one of them. 

```
018-09-23 12:36:21.273 [INF] RPCC: Block count: 33624
2018-09-23 12:36:26.263 [INF] RPCC: Balancer: Failed to connect to localhost:19110: dial tcp [::1]:19110: connect: connection refused
2018-09-23 12:36:26.263 [INF] RPCC: Connection state update for host localhost:19110: IDLE
2018-09-23 12:36:26.265 [INF] RPCC: Balancer: Failed to connect to localhost:19111: dial tcp [::1]:19111: connect: connection refused
2018-09-23 12:36:26.265 [INF] RPCC: Connection state update for host localhost:19111: IDLE
2018-09-23 12:36:26.265 [INF] RPCC: Balancer: Connection pick for RPC localhost:19109
2018-09-23 12:36:26.266 [INF] RPCC: Block count: 33624
2018-09-23 12:36:31.262 [INF] RPCC: Balancer: Failed to connect to localhost:19110: dial tcp [::1]:19110: connect: connection refused
2018-09-23 12:36:31.262 [INF] RPCC: Connection state update for host localhost:19110: IDLE
2018-09-23 12:36:31.264 [INF] RPCC: Balancer: Failed to connect to localhost:19111: dial tcp [::1]:19111: connect: connection refused
2018-09-23 12:36:31.264 [INF] RPCC: Connection state update for host localhost:19111: IDLE
2018-09-23 12:36:31.265 [INF] RPCC: Balancer: Connection pick for RPC localhost:19109
2018-09-23 12:36:31.265 [INF] RPCC: Block count: 33624
2018-09-23 12:36:36.261 [INF] RPCC: Balancer: Failed to connect to localhost:19110: dial tcp [::1]:19110: connect: connection refused
2018-09-23 12:36:36.261 [INF] RPCC: Connection state update for host localhost:19110: IDLE
2018-09-23 12:36:36.263 [INF] RPCC: Balancer: Failed to connect to localhost:19111: dial tcp [::1]:19111: connect: connection refused
2018-09-23 12:36:36.263 [INF] RPCC: Connection state update for host localhost:19111: IDLE
2018-09-23 12:36:36.263 [INF] RPCC: Balancer: Connection pick for RPC localhost:19109
2018-09-23 12:36:36.264 [INF] RPCC: Block count: 33624
2018-09-23 12:36:41.263 [INF] RPCC: Balancer: Failed to connect to localhost:19110: dial tcp [::1]:19110: connect: connection refused
2018-09-23 12:36:41.264 [INF] RPCC: Connection state update for host localhost:19110: IDLE
2018-09-23 12:36:41.266 [INF] RPCC: Balancer: Failed to connect to localhost:19111: dial tcp [::1]:19111: connect: connection refused
2018-09-23 12:36:41.266 [INF] RPCC: Connection state update for host localhost:19111: IDLE
2018-09-23 12:36:41.266 [INF] RPCC: Balancer: Connection pick for RPC localhost:19109
2018-09-23 12:36:41.266 [INF] RPCC: Block count: 33624
2018-09-23 12:36:46.264 [INF] RPCC: Balancer: Failed to connect to localhost:19110: dial tcp [::1]:19110: connect: connection refused
2018-09-23 12:36:46.264 [INF] RPCC: Connection state update for host localhost:19110: IDLE
2018-09-23 12:36:46.266 [INF] RPCC: Balancer: Failed to connect to localhost:19111: dial tcp [::1]:19111: connect: connection refused
2018-09-23 12:36:46.266 [INF] RPCC: Connection state update for host localhost:19111: IDLE
2018-09-23 12:36:46.266 [INF] RPCC: Balancer: Connection pick for RPC localhost:19109
2018-09-23 12:36:46.267 [INF] RPCC: Block count: 33624
2018-09-23 12:36:51.258 [INF] RPCC: Balancer: Connection pick for RPC localhost:19109
2018-09-23 12:36:51.259 [INF] RPCC: Block count: 33624
2018-09-23 12:36:56.257 [INF] RPCC: Balancer: Connection pick for RPC localhost:19109
...
```

This exhausts the retries for the other hosts if connections are not established. The problem with this is once the connection being used drops the balancer does not retry the other connections since their retry counts have already been maxed out.

```
2018-09-23 12:44:11.273 [INF] RPCC: Balancer: Connection pick for RPC localhost:19109
2018-09-23 12:44:11.274 [INF] RPCC: Block count: 33625
2018-09-23 12:44:16.274 [INF] RPCC: Balancer: Connection pick for RPC localhost:19109
2018-09-23 12:44:16.276 [INF] RPCC: Block count: 33625
2018-09-23 12:44:21.273 [INF] RPCC: Balancer: Connection pick for RPC localhost:19109
2018-09-23 12:44:21.275 [INF] RPCC: Block count: 33625
2018-09-23 12:44:26.275 [INF] RPCC: Balancer: Connection pick for RPC localhost:19109
2018-09-23 12:44:26.277 [INF] RPCC: Block count: 33625
2018-09-23 12:44:31.277 [INF] RPCC: Balancer: Connection pick for RPC localhost:19109
2018-09-23 12:44:31.279 [INF] RPCC: Block count: 33625
2018-09-23 12:44:36.273 [INF] RPCC: Balancer: Connection pick for RPC localhost:19109
2018-09-23 12:44:36.284 [INF] RPCC: Block count: 33625
2018-09-23 12:44:41.275 [INF] RPCC: Balancer: Connection pick for RPC localhost:19109
2018-09-23 12:44:41.281 [INF] RPCC: Block count: 33625
2018-09-23 12:44:46.274 [INF] RPCC: Balancer: Connection pick for RPC localhost:19109
2018-09-23 12:44:46.275 [INF] RPCC: Block count: 33625
2018-09-23 12:44:51.276 [INF] RPCC: Balancer: Connection pick for RPC localhost:19109
2018-09-23 12:44:51.287 [INF] RPCC: Block count: 33625
2018-09-23 12:44:56.274 [INF] RPCC: Balancer: Connection pick for RPC localhost:19109
2018-09-23 12:44:56.277 [INF] RPCC: Block count: 33625
2018-09-23 12:45:01.277 [INF] RPCC: Balancer: Connection pick for RPC localhost:19109
2018-09-23 12:45:01.286 [INF] RPCC: Block count: 33625
2018-09-23 12:45:06.277 [INF] RPCC: Balancer: Connection pick for RPC localhost:19109
2018-09-23 12:45:06.278 [INF] RPCC: Block count: 33625
2018-09-23 12:45:11.286 [INF] RPCC: Balancer: Connection pick for RPC localhost:19109
2018-09-23 12:45:11.287 [INF] RPCC: Block count: 33625
2018-09-23 12:45:13.470 [ERR] RPCC: Websocket receive error from localhost:19109: websocket: close 1006 (abnormal closure): unexpected EOF
2018-09-23 12:45:13.472 [INF] RPCC: Connection state update for host localhost:19109: DISCONNECTED
```

dnldd commented on rpcclient/balancer.go:
```diff
@@ -0,0 +1,67 @@
+	// for a hostaddress changes.
+	NotifyConnStateChange(sc *HostAddress, state ConnectionState, sync bool)
+
+	// UpdateReconnectAttempt will increase retryattempt + 1.
```
You're referring to a variable name that does not exist here: `retryattempt`, you probably meant `retryCount`. I'd suggest not using variable names in comments, it's easy to lose track of outdated comments when var names are modified. Plus a comment like `// UpdateReconnectAttempt increments the connection's retry counter by one.`  reads better imo.

dnldd commented on rpcclient/infrastructure.go:
```diff
@@ -153,12 +142,19 @@ type Client struct {
+	// balancer holds instance of RoundRobinBalancer. This implements
+	// the logic to round-robin between available connections.
+	// Whenever we need a connection, RoundRobinBalancer.NextConn() should be used to
+	// retrive a websocket connection or a HostAddress in case of HttpPostMode.
```
s/retrive/retrieve

maicalal commented:
Thanks @dnldd,
Have resolved all changes as suggested by @davecgh with checkin https://github.com/decred/dcrd/pull/1078/commits/0e73dafd1f0d3302341f6340cb11b32ee85a9e30

For the disconnect issue, there is a bug in the disconnect-reconnect workflow.
Am working on it.

maicalal commented:
Hi @dnldd , have resolved all previous comments by @davecgh .
Also have fixed the issues you faced.
The flow is as below:
Say con1 and con2 are specified to be used by balancer.
Initial connection (i.e. first successful connection):
- code would make 50 attempts for each connection, then mark it as shutdown.
Reconnection:
- after a successful initial connection, if either one of con1 and con2, disconnects,
this would result into making reconnection attempts as per the settings.
- during this phase, if there are no active connections, code will not execute any sendcmd (this is same as earlier, only difference is, earlier it would have used state of "disconnect" channel, but now we would need to rely on available active connections)

dnldd commented:
noted, I plan on reviewing soon. Will keep you updated.

dnldd commented:
here is the config I tested with:
```go
	connCfg := &rpcclient.ConnConfig{
		User:                 "user",
		Pass:                 "pass,
		DisableAutoReconnect: false,
		Certificates:         certs,
		HostAddresses: []rpcclient.HostAddress{
			rpcclient.HostAddress{
				Host:     "127.0.0.1:19110",
				Endpoint: "ws",
			},
			rpcclient.HostAddress{
				Host:     "127.0.0.1:19111",
				Endpoint: "ws",
			},
			rpcclient.HostAddress{
				Host:     "127.0.0.1:19109",
				Endpoint: "ws",
			},
		},
```
I initially started with `:19109(live), :19110(shutdown) and 19111(shutdown)`. I then killed `:19109` and started `:19110`, resulting in `:19109(shutdown), :19110(live) and 19111(shutdown)`. Here is the output for that:

```shell
2018-10-23 20:53:16.248 [INF] RPCC: Connection state update for host 127.0.0.1:19110: CONNECTING
2018-10-23 20:53:16.250 [INF] RPCC: Connection state update for host 127.0.0.1:19110: IDLE
2018-10-23 20:53:16.250 [INF] RPCC: Balancer: Failed to make initial connection to 127.0.0.1:19110: dial tcp 127.0.0.1:19110: connect: connection refused, attempt:1
2018-10-23 20:53:16.250 [INF] RPCC: Connection state update for host 127.0.0.1:19111: CONNECTING
2018-10-23 20:53:16.250 [INF] RPCC: Connection state update for host 127.0.0.1:19111: IDLE
2018-10-23 20:53:16.250 [INF] RPCC: Balancer: Failed to make initial connection to 127.0.0.1:19111: dial tcp 127.0.0.1:19111: connect: connection refused, attempt:1
2018-10-23 20:53:16.250 [INF] RPCC: Connection state update for host 127.0.0.1:19109: CONNECTING
2018-10-23 20:53:16.283 [INF] RPCC: Connection state update for host 127.0.0.1:19109: READY
2018-10-23 20:53:16.283 [INF] RPCC: Balancer: Established connection to RPC server 127.0.0.1:19109
2018-10-23 20:53:16.283 [INF] RPCC: Balancer: Connection pick for RPC 127.0.0.1:19109
2018-10-23 20:53:16.284 [INF] RPCC: Connection state update for host 127.0.0.1:19110: IDLE
2018-10-23 20:53:16.284 [INF] RPCC: Balancer: Failed to make initial connection to 127.0.0.1:19110: dial tcp 127.0.0.1:19110: connect: connection refused, attempt:2
2018-10-23 20:53:16.284 [INF] RPCC: Connection state update for host 127.0.0.1:19111: IDLE
2018-10-23 20:53:16.284 [INF] RPCC: Balancer: Failed to make initial connection to 127.0.0.1:19111: dial tcp 127.0.0.1:19111: connect: connection refused, attempt:2
2018-10-23 20:53:16.284 [INF] RPCC: Balancer: Connection pick for RPC 127.0.0.1:19109
2018-10-23 20:53:16.284 [INF] RPCC: Balancer: Will be using 127.0.0.1:19109 for notification: notifyblocks
2018-10-23 20:53:16.287 [INF] RPCC: NotifyBlocks: Registration Complete
2018-10-23 20:53:21.293 [INF] RPCC: Connection state update for host 127.0.0.1:19110: IDLE
2018-10-23 20:53:21.293 [INF] RPCC: Balancer: Failed to make initial connection to 127.0.0.1:19110: dial tcp 127.0.0.1:19110: connect: connection refused, attempt:3
2018-10-23 20:53:21.293 [INF] RPCC: Connection state update for host 127.0.0.1:19111: IDLE
2018-10-23 20:53:21.293 [INF] RPCC: Balancer: Failed to make initial connection to 127.0.0.1:19111: dial tcp 127.0.0.1:19111: connect: connection refused, attempt:3
2018-10-23 20:53:21.293 [INF] RPCC: Balancer: Connection pick for RPC 127.0.0.1:19109
2018-10-23 20:53:21.294 [INF] RPCC: Block count: 55353
2018-10-23 20:53:26.289 [INF] RPCC: Connection state update for host 127.0.0.1:19110: IDLE
2018-10-23 20:53:26.289 [INF] RPCC: Balancer: Failed to make initial connection to 127.0.0.1:19110: dial tcp 127.0.0.1:19110: connect: connection refused, attempt:4
2018-10-23 20:53:26.289 [INF] RPCC: Connection state update for host 127.0.0.1:19111: IDLE
2018-10-23 20:53:26.289 [INF] RPCC: Balancer: Failed to make initial connection to 127.0.0.1:19111: dial tcp 127.0.0.1:19111: connect: connection refused, attempt:4
2018-10-23 20:53:26.289 [INF] RPCC: Balancer: Connection pick for RPC 127.0.0.1:19109
2018-10-23 20:53:26.290 [INF] RPCC: Block count: 55353
2018-10-23 20:53:31.292 [INF] RPCC: Connection state update for host 127.0.0.1:19110: IDLE
2018-10-23 20:53:31.292 [INF] RPCC: Balancer: Failed to make initial connection to 127.0.0.1:19110: dial tcp 127.0.0.1:19110: connect: connection refused, attempt:5
2018-10-23 20:53:31.292 [INF] RPCC: Connection state update for host 127.0.0.1:19111: IDLE
2018-10-23 20:53:31.292 [INF] RPCC: Balancer: Failed to make initial connection to 127.0.0.1:19111: dial tcp 127.0.0.1:19111: connect: connection refused, attempt:5
2018-10-23 20:53:31.293 [INF] RPCC: Balancer: Connection pick for RPC 127.0.0.1:19109
2018-10-23 20:53:31.293 [INF] RPCC: Block count: 55353
2018-10-23 20:53:36.290 [INF] RPCC: Connection state update for host 127.0.0.1:19110: IDLE
2018-10-23 20:53:36.290 [INF] RPCC: Balancer: Failed to make initial connection to 127.0.0.1:19110: dial tcp 127.0.0.1:19110: connect: connection refused, attempt:6
2018-10-23 20:53:36.291 [INF] RPCC: Connection state update for host 127.0.0.1:19111: IDLE
2018-10-23 20:53:36.291 [INF] RPCC: Balancer: Failed to make initial connection to 127.0.0.1:19111: dial tcp 127.0.0.1:19111: connect: connection refused, attempt:6
2018-10-23 20:53:36.291 [INF] RPCC: Balancer: Connection pick for RPC 127.0.0.1:19109
2018-10-23 20:53:36.291 [INF] RPCC: Block count: 55353
2018-10-23 20:53:41.294 [INF] RPCC: Connection state update for host 127.0.0.1:19110: IDLE
2018-10-23 20:53:41.294 [INF] RPCC: Balancer: Failed to make initial connection to 127.0.0.1:19110: dial tcp 127.0.0.1:19110: connect: connection refused, attempt:7
2018-10-23 20:53:41.294 [INF] RPCC: Connection state update for host 127.0.0.1:19111: IDLE
2018-10-23 20:53:41.294 [INF] RPCC: Balancer: Failed to make initial connection to 127.0.0.1:19111: dial tcp 127.0.0.1:19111: connect: connection refused, attempt:7
2018-10-23 20:53:41.295 [INF] RPCC: Balancer: Connection pick for RPC 127.0.0.1:19109
2018-10-23 20:53:41.295 [INF] RPCC: Block count: 55353
2018-10-23 20:53:46.289 [INF] RPCC: Connection state update for host 127.0.0.1:19110: IDLE
2018-10-23 20:53:46.289 [INF] RPCC: Balancer: Failed to make initial connection to 127.0.0.1:19110: dial tcp 127.0.0.1:19110: connect: connection refused, attempt:8
2018-10-23 20:53:46.289 [INF] RPCC: Connection state update for host 127.0.0.1:19111: IDLE
2018-10-23 20:53:46.289 [INF] RPCC: Balancer: Failed to make initial connection to 127.0.0.1:19111: dial tcp 127.0.0.1:19111: connect: connection refused, attempt:8
2018-10-23 20:53:46.289 [INF] RPCC: Balancer: Connection pick for RPC 127.0.0.1:19109
2018-10-23 20:53:46.290 [INF] RPCC: Block count: 55353
2018-10-23 20:53:50.696 [ERR] RPCC: Websocket receive error from 127.0.0.1:19109: websocket: close 1006 (abnormal closure): unexpected EOF
2018-10-23 20:53:50.697 [INF] RPCC: Connection state update for host 127.0.0.1:19109: DISCONNECTED
2018-10-23 20:53:50.698 [INF] RPCC: Failed to connect to 127.0.0.1:19109: dial tcp 127.0.0.1:19109: connect: connection refused
2018-10-23 20:53:50.698 [INF] RPCC: Retrying connection to 127.0.0.1:19109 in 10s
2018-10-23 20:54:00.701 [INF] RPCC: Failed to connect to 127.0.0.1:19109: dial tcp 127.0.0.1:19109: connect: connection refused
2018-10-23 20:54:00.701 [INF] RPCC: Retrying connection to 127.0.0.1:19109 in 15s
2018-10-23 20:54:15.702 [INF] RPCC: Failed to connect to 127.0.0.1:19109: dial tcp 127.0.0.1:19109: connect: connection refused
2018-10-23 20:54:15.702 [INF] RPCC: Retrying connection to 127.0.0.1:19109 in 20s
2018-10-23 20:54:35.708 [INF] RPCC: Failed to connect to 127.0.0.1:19109: dial tcp 127.0.0.1:19109: connect: connection refused
2018-10-23 20:54:35.708 [INF] RPCC: Retrying connection to 127.0.0.1:19109 in 25s
2018-10-23 20:55:00.713 [INF] RPCC: Failed to connect to 127.0.0.1:19109: dial tcp 127.0.0.1:19109: connect: connection refused
2018-10-23 20:55:00.713 [INF] RPCC: Retrying connection to 127.0.0.1:19109 in 30s
2018-10-23 20:55:30.719 [INF] RPCC: Failed to connect to 127.0.0.1:19109: dial tcp 127.0.0.1:19109: connect: connection refused
2018-10-23 20:55:30.719 [INF] RPCC: Retrying connection to 127.0.0.1:19109 in 35s
2018-10-23 20:56:05.725 [INF] RPCC: Failed to connect to 127.0.0.1:19109: dial tcp 127.0.0.1:19109: connect: connection refused
2018-10-23 20:56:05.725 [INF] RPCC: Retrying connection to 127.0.0.1:19109 in 40s
2018-10-23 20:56:45.728 [INF] RPCC: Failed to connect to 127.0.0.1:19109: dial tcp 127.0.0.1:19109: connect: connection refused
2018-10-23 20:56:45.728 [INF] RPCC: Retrying connection to 127.0.0.1:19109 in 45s
2018-10-23 20:57:30.734 [INF] RPCC: Failed to connect to 127.0.0.1:19109: dial tcp 127.0.0.1:19109: connect: connection refused
2018-10-23 20:57:30.734 [INF] RPCC: Retrying connection to 127.0.0.1:19109 in 50s
...
```

From the logs, it's continuously retrying `:19109` when at this point it should be trying to connect to  `:19109, :19110 and :19111`  like it did on startup. On a disconnect cycle though all  host addresses for a connection, if none is found you back off the retry cycle eg. `Retrying all host addresses in 10s, 20s, 30s, 40s ...` 

I'm yet to peruse you updates on the comments, expect a review on that soon.

maicalal commented:
Hi @dnldd , 
The reconnect handler kicks in once a connection gets disconnected. As we can see in the logs
`Retrying connection to 127.0.0.1:19109 in ...s`
The connections for :19110 and :19111 never got connected, so they do not get into the reconnect workflow.

Do you want me to change the reconnect handler to work on these as well ?
This will be a small change. Didn't do it earlier because I thought they do not qualify, as they never got connected.

dnldd commented:
Yeah, let's do that. On a disconnection retry all host addresses.

maicalal commented:
hi @dnldd , have updated the code accordingly. 
Please try with latest.

dnldd commented on rpcclient/balancer.go:
```diff
@@ -0,0 +1,67 @@
+
+	// NotifyReconnect will update the map for wsConns and update the connection state
+	// for this address.
+	NotifyReconnect(wsConn *websocket.Conn, hostAdd *HostAddress)
```
NotifyReconnect updates the connection and state of the provided host address on reconnection.

dnldd commented on rpcclient/balancer.go:
```diff
@@ -0,0 +1,67 @@
+
+	// Close closes the passed ws connection and sets the state as Disconnected
+	// by calling NotifyConnStateChange.
+	// Returns true if this was the last WS connection to be closed.
```
// Close terminates the passed connection and updates its state to disconnected
// via NotifyConnStateChange. It returns true if the connection was the last active 
// connection.

dnldd commented on rpcclient/balancer.go:
```diff
@@ -0,0 +1,67 @@
+	// Returns true if this was the last WS connection to be closed.
+	Close(wsConn *websocket.Conn) bool
+
+	// CloseAll closes all the ws connections which are active at present.
```
... terminates all active connections.

dnldd commented:
When all hosts are unavailable, the initial retry cycle does not back off.

```2018-11-13 02:21:26.132 [INF] RPCC: Connection state update for host 127.0.0.1:9110: CONNECTING
2018-11-13 02:21:26.137 [INF] RPCC: Connection state update for host 127.0.0.1:9110: IDLE
2018-11-13 02:21:26.137 [INF] RPCC: Balancer: Failed to make initial connection to 127.0.0.1:9110: dial tcp 127.0.0.1:9110: connect: connection refused, attempt:1
2018-11-13 02:21:26.137 [INF] RPCC: Connection state update for host 127.0.0.1:9111: CONNECTING
2018-11-13 02:21:26.138 [INF] RPCC: Connection state update for host 127.0.0.1:9111: IDLE
2018-11-13 02:21:26.138 [INF] RPCC: Balancer: Failed to make initial connection to 127.0.0.1:9111: dial tcp 127.0.0.1:9111: connect: connection refused, attempt:1
2018-11-13 02:21:26.138 [INF] RPCC: Connection state update for host 127.0.0.1:9109: CONNECTING
2018-11-13 02:21:26.139 [INF] RPCC: Connection state update for host 127.0.0.1:9109: IDLE
2018-11-13 02:21:26.139 [INF] RPCC: Balancer: Failed to make initial connection to 127.0.0.1:9109: dial tcp 127.0.0.1:9109: connect: connection refused, attempt:1
2018-11-13 02:21:26.139 [INF] RPCC: Connection state update for host 127.0.0.1:9110: IDLE
2018-11-13 02:21:26.149 [INF] RPCC: Balancer: Failed to make initial connection to 127.0.0.1:9110: dial tcp 127.0.0.1:9110: connect: connection refused, attempt:2
2018-11-13 02:21:26.151 [INF] RPCC: Connection state update for host 127.0.0.1:9111: IDLE
2018-11-13 02:21:26.151 [INF] RPCC: Balancer: Failed to make initial connection to 127.0.0.1:9111: dial tcp 127.0.0.1:9111: connect: connection refused, attempt:2
...
2018-11-13 02:21:27.052 [INF] RPCC: Connection state update for host 127.0.0.1:9110: IDLE
2018-11-13 02:21:27.052 [INF] RPCC: Balancer: Failed to make initial connection to 127.0.0.1:9110: dial tcp 127.0.0.1:9110: connect: connection refused, attempt:47
2018-11-13 02:21:27.054 [INF] RPCC: Connection state update for host 127.0.0.1:9111: IDLE
2018-11-13 02:21:27.089 [INF] RPCC: Balancer: Failed to make initial connection to 127.0.0.1:9111: dial tcp 127.0.0.1:9111: connect: connection refused, attempt:47
2018-11-13 02:21:27.090 [INF] RPCC: Connection state update for host 127.0.0.1:9109: IDLE
2018-11-13 02:21:27.090 [INF] RPCC: Balancer: Failed to make initial connection to 127.0.0.1:9109: dial tcp 127.0.0.1:9109: connect: connection refused, attempt:47
2018-11-13 02:21:27.091 [INF] RPCC: Connection state update for host 127.0.0.1:9110: IDLE
2018-11-13 02:21:27.091 [INF] RPCC: Balancer: Failed to make initial connection to 127.0.0.1:9110: dial tcp 127.0.0.1:9110: connect: connection refused, attempt:48
2018-11-13 02:21:27.091 [INF] RPCC: Connection state update for host 127.0.0.1:9111: IDLE
2018-11-13 02:21:27.091 [INF] RPCC: Balancer: Failed to make initial connection to 127.0.0.1:9111: dial tcp 127.0.0.1:9111: connect: connection refused, attempt:48
2018-11-13 02:21:27.092 [INF] RPCC: Connection state update for host 127.0.0.1:9109: IDLE
2018-11-13 02:21:27.113 [INF] RPCC: Balancer: Failed to make initial connection to 127.0.0.1:9109: dial tcp 127.0.0.1:9109: connect: connection refused, attempt:48
...
```

50 retries for each host are exhausted in a matter of seconds as a result. Is there any reason why you're not backing off here as compared to a dropped connection? Here is an example:
```
...
2018-11-13 02:26:24.092 [INF] RPCC: Retrying connection to 127.0.0.1:19110 in 10s
2018-11-13 02:26:34.095 [INF] RPCC: Failed to connect to 127.0.0.1:19111: dial tcp 127.0.0.1:19111: connect: connection refused
2018-11-13 02:26:34.095 [INF] RPCC: Retrying connection to 127.0.0.1:19111 in 10s
2018-11-13 02:26:44.101 [INF] RPCC: Failed to connect to 127.0.0.1:19109: dial tcp 127.0.0.1:19109: connect: connection refused
2018-11-13 02:26:44.101 [INF] RPCC: Retrying connection to 127.0.0.1:19109 in 10s
2018-11-13 02:26:54.104 [INF] RPCC: Failed to connect to 127.0.0.1:19110: dial tcp 127.0.0.1:19110: connect: connection refused
2018-11-13 02:26:54.104 [INF] RPCC: Retrying connection to 127.0.0.1:19110 in 15s
...
```

maicalal commented:
Hi @dnldd , I assumed those which never got connected, need to be handled differently.
Looks like am wrong here.
Will fix the code to match the backoff strategy as in the case of dropped connections.
Working on it.

maicalal commented:
Hi @dnldd have incorporated your suggestions. Please try with latest.

maicalal commented:
Hi @dnldd, please let me know if you want me to make any more changes...

dajohi commented on rpcclient/roundrobin.go:
```diff
@@ -0,0 +1,488 @@
+		}
+	}
+	if conn == nil {
+		return conn, ErrNoUsableConnAvailable
```
missing Unlock

maicalal replied:
fixing it.

dajohi commented on rpcclient/roundrobin.go:
```diff
@@ -0,0 +1,488 @@
+					"%s", hostAddress.Host, scaledDuration)
+				time.Sleep(scaledDuration)
+			}
+			wsConn, err = dial(rrb.connConfig)
```
should move this outside the lock

maicalal replied:
fixing it.

jrick commented:
FYI - I have no plans on pulling this change into dcrwallet due to my aforementioned concerns of its design.  Having multiple RPC servers to sync the wallet with is useful, but the syncer should have direct knowledge of each of these severs (similar to how the SPV syncer directly handles all connected peers).

I was already planning on removing the rpcclient dependency from dcrwallet entirely for other reasons (it's already more complex to work with than it should be), but this would just accelerate that process.

maicalal commented:
> have you built and ran this with `go build -race`?

Hi @dajohi, had run with -race but it did not detect the code you pointed out in picker.
Maybe because it goes into this if block only once, then it errors out saying no connection available.
Will checkin the changes in sometime.

davecgh commented:
This never really got any traction, since, as it turns out, all primary consumers wanted to control the connection management themselves.  One of them in particular, the wallet, decided to move away from rpcclient altogether.  Consequently, while I really am not happy about closing this without merging after all this effort and time, none of the consumers want it, and it would just end up being code that has to be maintained for no benefit.  Closing.
